### ISSUE TITLE
Quality of Life Pass/Pipeline Features

### ISSUE THREAD

**Issue by AlexanderViand-Intel** [google/heir#1107]
Small fixes/add-ons that I noticed while drawing up a pass/pipeline overview figure:

- [x] `--secretize`'s `entry-function` parameter should be optional and, if omitted, the pass should apply to all functions in the input (module?). This should then also be reflected in any pipelines that use `--secretize`, making their `entry-function` optional, too.
- [x]  In order to better support mixed ctxt/ptxt examples, and for better integration with frontends, create `--annotated-mlir-to-....` that are the same as the current `--mlir-to-...` pipelines but without the `--secretize` pass.
- [ ] The `ciphertext-degree` parameter in the various passes & pipelines that use it should be optional, with the compiler using the size of (the right-most-non-one-dimension of) the first secret tensor in the function signature if the parameter is omitted (and erroring out if no such secret tensor function parameter exists)
See: #1197
- [x] `--optimize-relinearization` should be added to the RLWE pipeline builder? (EDIT: Not until #1148, and ideally also #1032 are fixed)
- [x] `--operation-balancer` shuld be added to the mlir-to-secret-arithmetic pipeline
- [x] Investigate if the `--bgv-to-openfhe` and `--ckks-to-openfhe` passes can be merged into a single `--rlwe-to-openfhe` pass combining the patterns from both passes.
Also: #1150
- [x] (Optional/TBD): Split the library-specific parts of the pipeline into separate pipelines. While that might require doing something like `--mlir-to-ckks --rlwe-to-openfhe`, it might be nicer to keep backend-specific stuff separate/modular. 
- [ ] Better error messages ( #1143)

**Comment by AlexanderViand-Intel** [comment]
Based on discussion from today's office hours: 

For the `ciphertext-degree` parameter, we should decouple that into the actual ctxt degree (which can be optional and default to something reasoanble such as 16k) and something like `--assume-native-tensor-size` that does what was mentioned above.

In general, these things should not be just stored in a single pass's paramaters but encoded into the IR (at the module level)

**Comment by ZenithalHourlyRate** [comment]
Also I'm wondering why `--operation-balancer` did not get into the `mlir-to-secret-arithmetic` pipeline, though its absense makes me happy when constructing deep circuit for debugging.

**Comment by ZenithalHourlyRate** [comment]
> Do you happen to know which tests you expect to break?

No, haven't tried it.

**Comment by AlexanderViand-Intel** [comment]
With #1196 addressing the last "clean up" part of this and the ciphertext degree thing + error messages being quite significant work packages in their own right with their own issues to track them, I've set this to close when #1196 gets merged.
